### ISSUE TITLE
Compilation error with Magma Library

### ISSUE THREAD

**Issue by Anshuman5** [dftbplus/dftbplus#514]
Dear Developers,

I get the following error while trying to build DFTB with GPU via Magma library.
Magma version is 2.5.3 and cuda/10.1.

dftbplus/_build/prog/dftb+/lib_magmahelper/magma_device_info.f90:27.24:

    subroutine  gpu_req8(req_ngpus) bind(c, name='MagmaGpuNum_Requested')
                        1
/rhome/haizhou/bigdata/anshuman/plumed_gpu/dftbplus/_build/prog/dftb+/lib_magmahelper/magma_device_info.f90:22.24:

    subroutine  gpu_req4(req_ngpus) bind(c, name='MagmaGpuNum_Requested')
                        2
Error: Binding label 'MagmaGpuNum_Requested' in interface body at (1) collides with the global entity 'MagmaGpuNum_Requested' at (2)
Assembler messages:
Internal error (Illegal instruction).
Please report this bug.
make[2]: *** [prog/dftb+/CMakeFiles/dftbplus.dir/lib_magmahelper/magma_device_info.f90.o] Error 1
make[2]: *** [prog/dftb+/CMakeFiles/dftbplus.dir/lib_io/fileid.f90.o] Error 1

Regards

**Comment by bhourahine** [comment]
Can you supply details of which C and Fortran compilers was used including version numbers (the error suggest it is an internal error produced from there)?

**Comment by Anshuman5** [comment]
C and Fortran compiler identification is GNU 4.8.5. When I change this to GNU 7.4.0 the above issue is resolved. Using gnu.cmake file (after including the path for plumed and Magma library) , I get the following error:
f951: Fatal Error: Reading module 'magma' at line 1 column 2: Unexpected EOF
compilation terminated.
make[2]: *** [prog/dftb+/CMakeFiles/dftbplus.dir/lib_math/eigensolver.f90.o] Error 1

I change from gnu.cmake to intel.cmake and load the intel/2018 modules. This compiles without any error.

**Comment by bhourahine** [comment]
Unfortunately the GNU 4.8.5 compiler is too old to support recent Fortran standards so is not supported by the code.

Did you re-compile MAGMA with 7.4.0, or was this built with the older compiler? Also did you fully remove the dftb+ build from the older compiler before starting?

Its good to hear that intel 18 compiles for you, have you tried testing the resulting binary with ctest ?

**Comment by Anshuman5** [comment]
I did recompile MAGMA with 7.4.0 and before starting a new build, fully remove the older build. I ran a geometry optimization and it is working. 
Another issue, I want to discuss is the performance of the DFTB code when compiled with both MAGMA and PLUMED libraries. (since GPU is not supported with MPI code, Plumed and DFTB are compiled without MPI support) 
It is much faster (depending on the system size) when MAGMA eigensolver is used compared to other solvers. 
For my research project, I need to run a metadynamics calculation, with ~1K atoms(# of H atoms=666,0=329,N-2,C=6)  1ns metadynamics would take ~83 days.
1ns ---> 10^6 fs ---> 10^6 geometry steps (each step size=1fs). Each geometry step takes ~7scc cycles (~7secs) to converge.

I want to understand if there is a way to improve the performance further? 

I am attaching the Dockerfile which I used to compile the DFTB-PLUMED-MAGMA on the cloud.
[Dockerfile.txt](https://github.com/dftbplus/dftbplus/files/4828260/Dockerfile.txt)
